### PR TITLE
Precomputed entity attributes

### PR DESCRIPTION
**Pull request by cdisselkoen**
[Rendered](https://github.com/cedar-policy/rfcs/blob/precomputed-entity-attributes/text/0034-precomputed-entity-attributes.md)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by andrewmwells-amazon** [comment]
For Component 2 I slightly prefer the 2nd option (consolidating `EvalResult` and `Value`), but I'm fine with the RFC as-is.

**Review by andrewmwells-amazon — APPROVED** [review]

**Review by shaobo-he-aws — APPROVED** [review]

**Review by anwarmamat — APPROVED** [review]

File: text/0034-precomputed-entity-attributes.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+return `EntitiesError` but not an evaluation error.
+- Conversely, `Entity::attr()` need not return an error anymore, as all of the
+attributes will already be evaluated. (If we are very concerned about backward
+compatibility, we could keep the current signature and just never return `Err`.
+Currently, this RFC is proposing we change the signature to not contain
+`Result`, since we're already making the related breaking change to
+`Entity::new()`.)
+
+Accepting these breaking changes allows us to give users the best-performing
+behavior by default.
```
Comment (mwhicks1) [inline text/0034-precomputed-entity-attributes.md:68]: What will happen to customer code when they upgraded? I think they will get a type error, which they can easily fix, is that right?

File: text/0034-precomputed-entity-attributes.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+- not expose `Value` and just use `EvalResult`, even in the new constructors.
+This provides maximum backwards compatibility, like the first option, and
+also avoids the possible confusion generated by the first option, but comes
+at a notable performance cost (due to the inefficient
+`EvalResult`-to-`Value` conversion required under the hood) and might
+generate a different kind of confusion as `EvalResult` is an awkward name in
+the context of these constructors.
+
+This RFC currently proposes we take the third option.
+
```
Comment (mwhicks1) [inline text/0034-precomputed-entity-attributes.md:108]: Same question as above: What happens when you try to compile existing code, which will use `EvalResult` to use the new API? I think you will get a type error that is easily fixed?

File: text/0034-precomputed-entity-attributes.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+In addition to everything proposed in the current RFC, we could provide the
+status-quo definitions of `Entities` and `Entity` under new names: e.g.,
+`UnevaluatedEntities` and `UnevaluatedEntity` (naming of course TBD).
+This is very unlikely to be of any use to new users, but existing users really
+wanting the old error behavior could use these types and functions instead of
+`Entities` and `Entity`.
+
+To me, this provides little benefit, at the expense of more confusion and more
+APIs to maintain. It also doesn't make any of the changes more back-compatible;
```
Comment (mwhicks1) [inline text/0034-precomputed-entity-attributes.md:160]: +1

File: text/0034-precomputed-entity-attributes.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+We could do only first and third components of this RFC, and not the second
+component, at least for now.
+(The second component -- adding the new constructors for `Entities` and `Entity`
+-- is severable and could be its own RFC, or perhaps a change that doesn't rise
+to the level of requiring an RFC.)
+
+In essence, we could decide that the second component is more trouble than it's
+worth -- that it provides only marginal benefit to users at the cost of
+cluttering our API, both with additional constructors, and with the public
+`Value` type.
```
Comment (mwhicks1) [inline text/0034-precomputed-entity-attributes.md:138]: I tend to agree with this assessment.

**Review by mwhicks1 — COMMENTED** [review]
I'm leaning toward Alternative A, but I'm curious about your response to my questions.

File: text/0034-precomputed-entity-attributes.md (hunk 1 above)
Comment (cdisselkoen) [reply text/0034-precomputed-entity-attributes.md:68]: Yes, users will get type errors (or other compile errors) for all the breaking changes contemplated in this RFC.  As far as "easily fix", existing users of `EvalResult` (if there are any) will have the most involved fixes, as the underlying structure of `EvalResult` will have changed (if we take the third option in Component 2, as currently proposed).  The fixes for the changes in Component 1 should be easy for users. (and Component 3 is not a breaking change for any users on a released Cedar version -- and even if someone is on a pre-release version of Cedar and is using `.evaluate()`, they simply have to remove it.)

File: text/0034-precomputed-entity-attributes.md (hunk 2 above)
Comment (cdisselkoen) [reply text/0034-precomputed-entity-attributes.md:108]: See my answer above.  You will get a compile error and have to adapt to the new structure, depending on how you are using the `EvalResult`.  I don't know if anyone is actually using `EvalResult`, and if so, how they are using it.  If they are just printing the result to a string/console (as the CLI does), then the fix is minimal or zero, as the new version will be equally formattable with the Rust `Display` trait.

File: text/0034-precomputed-entity-attributes.md (hunk 4 above)
Comment (cdisselkoen) [reply text/0034-precomputed-entity-attributes.md]: Updated this RFC to move the second component to Alternative A (and adopt the previous Alternative A as the main proposal).  See my comment below.

File: text/0034-precomputed-entity-attributes.md (hunk 2 above)
Comment (cdisselkoen) [reply text/0034-precomputed-entity-attributes.md]: Resolving this thread now that this component is no longer in the main proposal.

File: text/0034-precomputed-entity-attributes.md (hunk 1 above)
Comment (cdisselkoen) [reply text/0034-precomputed-entity-attributes.md:64]: Now that the RFC has been updated to remove Component 2 (leaving it in Alternative A), the only breaking changes that remain are the ones in Component 1, which should be easy for users (and yes, surface as type errors or other compile errors).

**Review by mwhicks1 — APPROVED** [review]
LGTM

**Review by khieta — APPROVED** [review]

**Pull request merged**